overlord/state/taskrunner.go

```diff
@@ -301,7 +333,7 @@ func (r *TaskRunner) run(t *Task) {
 				r.taskErrorCallback(err)
 			}
 		}
-
+		r.checkExhaustion(t.Change())
```

Comment:
I wonder if there are situations where a Change status can change without involving the return from a task handler, and whether we should unify this/do something closer to taskStatusChanged and what @mvo is doing in #12898?

Reply:
I went through our uses of Task.SetStatus(), and in general it seems like our code does not manipulate the explicit statuses of tasks, other than the task-based handler code, where they usually only manipulate their own status (which wouldn't be an issue). I did however spot a usage in aliasesv2.go (line 473), where this is not true. Theoretically here this could be an issue.

I took a look at @mvo5 solution and I like the idea of hooking onto tasks/changes specifically instead of using the taskrunner. However we would probably need an identical hook just for change-level hooks instead of a task-based one.

So maybe:

```
// AddTaskStatusChangedObserver adds the given callback to the list of
// callbacks that are called when a task changes it's status.
func (s *State) AddChangeStatusChangedObserver(cb func(chg *Change, old, new Status)) {
	s.writing()
	s.changeStatusChangedObservers = append(s.changeStatusChangedObservers, cb)
}

func (s *State) notifyChangeStatusChangedObservers(chg *Change, old, new Status) {
	for _, f := range s.changeStatusChangedObservers {
		// XXX: this must not block or we cannot run changes anymore
		// XXX2: stop running callbacks when there is an error returned
		f(chg, old, new)
	}
}
```

Reply:
I can try to use this approach here and keeping more state in taskrunner in #12898 - I like callbacks in the taskrunner better. It seems we are doing SetStatus() in handlers consistently and then return (or schedule a reboot) but I need to poke some more if I miss corner-cases.